### DISCUSSION TITLE
Globe on zoom out via Adaptive Composite Map Projection

### DISCUSSION THREAD

**Discussion by klokan** [maplibre/maplibre#161] [Cross-project feature proposals]
![Snímek_obrazovky_2021-06-11_v 18 15 30](https://user-images.githubusercontent.com/59284/207892881-c91294f5-7ea9-4003-a8b8-cb98a0e70841.png)

Ability to show and interact with Globe (or alternative view on earth) when map is zoomed out - while reusing and loading the same Mercator vector tile and raster data - client side reprojection of the data - using Adaptive Composite Map Projection - where deeply zoomed you are in Mercator - but from certain zoom level up there is transition to another projection. 

Similar functionality as Google Maps or Mapbox has now.

The scientific articles:
https://berniejenny.info/demos/AdaptiveCompositeMapProjections/
https://github.com/OSUCartography/adaptive-composite-map-projections
http://cartography.oregonstate.edu/demos/AdaptiveCompositeMapProjections/

also 

https://openlayers.org/doc/tutorials/raster-reprojection.html (developed by MapTiler team in OpenLayers)

or

https://observablehq.com/@fil/webmercator-to-globe?collection=@fil/glsl-projections  

and

https://www.jasondavies.com/maps/transition/
https://www.jasondavies.com/maps/tile/  
https://www.jasondavies.com/maps/rotate/

and

![google-maps-globe](https://user-images.githubusercontent.com/59284/208049031-3d5494c9-acf3-425a-ab3c-bacc8f20c1c2.gif)

**Ideas / tickets towards this direction should be proposed in different CORE projects of MapLibre.**

Ported from https://roadmap.maplibre.org/c/92-globe-on-zoom-out-via-adaptive-composite-map-projection

**Comment by lseelenbinder** [comment]
I would love to see this, as well! It really adds a natural feel to the map.

**Comment by wipfli** [comment]
This project was selected to become a Bounty Direction. You find a tracking issue at https://github.com/maplibre/maplibre/issues/190

**Comment by Dunrar** [comment]
I also love the idea of different projections as well as a globe view. But I do not see how exactly this is related to the Adaptive Composite Map Projections paper. 

As far as I can see, the paper describes a combination of sane defaults, techniques and tricks to adapt/combine well known map projections with desirable properties, based on zoom level, lat/long and different user requirements. But for very small scales it deals only with projections that show the whole earth, not just one hemisphere (in contrast to globe-like projections like the orthographic or all kinds of general perspective projections). The technique also sacrifices continuity if deemed necessary to retain some other properties (most of all equal area and also a straight equator when centered on etc.). Actually slight discontinuity even happens in the transition between the lambert azimuthal and  albers conic projections, which is the centerpiece of the technique. 

To me, this is not what the Google Maps example looks like. It looks much more like a simple, continuous interpolation between mercator and orthographic. Examples can be seen at  https://observablehq.com/@jashkenas/interpolating-d3-map-projections and https://observablehq.com/@cguastini/mercator-to-orthographic. This should be much, much easier to implement than ACMP or even just all the projections ACMP is using, right? Please correct me if I overlooked or misunderstood something.

**Comment by birkskyum** [comment]
Does it make sense to break this into two efforts - Globe Mode (no mercator on zoom) and Adaptive Projections (2D only, for now), and then make then work together after?

Is there some shared work that would have to be done first in that case, that we can be more clear about?

**Comment by birkskyum** [comment]
Here is an interesting writeup of the challenges involved in implementing this https://tangrams.readthedocs.io/en/main/Overviews/Projections-Overview/

**Reply by sjg-wdw** [reply]
You could almost pull this off with modern shaders.  The details kind of make my head hurt, though.

**Comment by sjg-wdw** [comment]
Hi, we're looking at the bounties and considering proposing a bundle of them together.  Most are either straightforward or vague enough that we can come up with a very good definition.  But this one is both very specific and not the way we'd do it.

This is 'we' as in Wet Dog Weather, not the Metal team.

So let me throw this one out.  We did multiple coordinate systems at the same time on a globe and we definitely did not do it this way.  It was more straightforward, but also more complex in its own way.

My question would be, are you focused on this approach or how do you feel about a slightly more mundane approach?

**Reply by sjg-wdw** [reply]
How did we do it?  Well, it's gets complicated, but it starts with Proj and a target coordinate system.  If we're displaying in Web Mercator, then tiles are reprojected to Web Mercator.  For a globe, we convert to geocentric and then scale to a model space with a radius of 1.0.  But it'll work with any coordinate system as a target, with the caveat that you do get weirdness in a lot of them (polar anyone?) and it's best to stick to a handful that work fairly logically.

The actual geometry conversion is not all that bad, though there are a lot of random issues.  Where it gets really hard is deciding what tiles to load when your source coordinate system is different from your display coordinate system.  We do have a solution, it is open source, and we'd be willing to move it over.

**Reply by sjg-wdw** [reply]
Then there's the gestures.  Another area where things get weird.  MapLibre is basically a semi-smart flat map.  It's not so dumb to work in pixels... but it kind of does in a lot of cases.  So you need to unkink a lot of logic that assumes a fairy simple mapping from screen space to geographic.... or at least the ability to go through geographic when it's convenient.

And don't forget 3D terrain.  Now you need to grab things that pop up, you need to take elevation into account for the viewer and you probably should be supporting an ellipsoid or some users will get very mad at you.  That means your globe isn't even a sphere.... it gets to be a lot.

**Reply by HarelM** [reply]
Personally speaking, I don't think someone has strong feelings towards a specific approach, I think everyone would love to just have it.
If there's a way to split the work to chunks it would be great.
I don't know if you meant for supporting this just in native but doing it web as well would be very much welcomed. 😀

**Reply by sjg-wdw** [reply]
If someone does have a strong opinion on this particular approach, I wanted to hear it before proposing something different.

In terms of which platform, we know native very well now, so it's easier to propose.  The same approach would likely work on the web, but we'd need to do some digging to come up with a proper estimate.

**Reply by HarelM** [reply]
Cool, thanks for the info.

**Reply by birkskyum** [reply]
@sjg-wdw , how does it compare to this?  obviously the tip for web requires the 3d terrain which isn't in Native yet. https://github.com/maplibre/maplibre-gl-js/issues/307#issuecomment-1547772875

**Reply by sjg-wdw** [reply]
There are some similarities to the way we do it in WhirlyGlobe.  Such as the way we handle the poles and how we render the image and fill polygons of tiles to an offscreen buffer.  But then we drape the lines on top by sort of chopping them up a bit.  This person has the right idea, though.

We do go all in, though.  So if you're in globe mode you're all the way in globe mode, down to the surface.  Causes a lot of work with gestures and picking and managing offsets to keep coordinates fairly small.  But it makes the aviation folks happy.

**Reply by sjg-wdw** [reply]
It looks like Petr at @klokan was maybe one of the early proponents of this Composite approach.  If he has some thoughts I'd love to hear them.

But I think what we'll do is propose the WhirlyGlobe approach as a mini whitepaper in the bounties and see what people think.

**Comment by mdorda** [comment]
Globe view is something we are really interested in here at Windy.com :-) We cannot promise any concrete result, but we would like to contribute as a company to the development of MapLibre. That is why we are trying to develop this functionality as part of a student project at Charles University in Prague. If all goes well, work should start in October. We will try to keep you updated.

**Reply by birkskyum** [reply]
Interesting. @mdorda I reckon that would be related exclusively to GL JS right?

**Reply by mdorda** [reply]
Yes, unfortunately only GL JS is relevant for us.

**Reply by birkskyum** [reply]
That is okay - we aim to keep the feature set in sync, but in practice it's different teams pushing our two renderers forwards, so a contribution to one of the renderers is still value, where we'll then try have the other renderer catch-up with similar functionality. We operate with Design proposals, which is the functional and technical [design decisions](https://github.com/maplibre/maplibre-native/tree/main/design-proposals). It's best to start out adding such a PR with such a proposal for feedback by the community (to the GL JS repo - there aren't a folder yet).

**Reply by wipfli** [reply]
Love windy

**Reply by mdorda** [reply]
See comment https://github.com/maplibre/maplibre/discussions/161#discussioncomment-7630374 from @kubapelc who is working on this with Windy. Also see @Pheonor is working hard on this. Could not we join forces somehow?

**Comment by birkskyum** [comment]
@klokan , It's come to my attention that Google Maps in Globe View does actually not go to a Mercator projection on high zoom (can be verified by zooming in and toggle globe view). Of course it should be possible to keep loading mercator data (and ideally data from any projection, which is a different effort), but given that it increases the complexity considerably, can you elaborate what value is to be found by rendering a Mercator projection on high zoom, rather than maintaining the globe projection?

**Reply by sjg-wdw** [reply]
I'll add that our aviation users for WhirlyGlobe were the pickiest we had about coordinate systems.  They cared about datums and projections and liked working on a globe as long as the data was properly reprojected.  If you got close to the surface, you needed to respect the ellipsoid.

**Comment by Pheonor** [comment]
Hi,

I want to share with you the proof of concept I made on MapLibre GL JS.

### Projections;
The first think I do, is to add a "projection" class to handle three type of projection (for the moment, probably the system could be extended with other projections):
- mercator: the default one
- globe: the 3D Globe view
- adaptative: an automatic transition between the mercator and globe projection

The globe view uses the same technique as the terrain with an offscreen pass then a 3D pass with the previous computed texture.
All the 3D globe projection is performed into the vertex shader. This system allows to switch between the two projections with the same shader. All UV coordinates are reprojected also to ensure good position of each textures.

![Capture d'écran 2023-11-19 133856](https://github.com/maplibre/maplibre/assets/2014432/a062ff82-cfc4-4424-b23a-1f4c4c4bbf40)

The adaptative projection made a linear transition from mercator and globe projection between two zoom levels.

https://github.com/maplibre/maplibre/assets/2014432/b84b4d77-76f2-46e5-810c-97efab7fa2e6

In practice, the linear projection is not the best, but main transformation is outside the camera view, so the simplification is not visible.

When starting the folding process at lower zoom level, the system could be observed.

https://github.com/maplibre/maplibre/assets/2014432/50ed0273-cdf7-4419-b06e-a3378d08ed44


### Tiles:
To ensure all tiles are loaded, I change the tile loading system to check not only the four corners of each tiles but also a grid inside each tile (the number of point is dynamically computed according to the zoom level) ensuring to load the tile if it is into the camera frustrum even on Earth border view.

### Atmosphere and stars:

To have a nice display when zoom out, I added a realistic atmosphere shaders.
This atmosphere hide progressively when zoom in to let the full earth displayed without altering colors.

To avoid a large black background, I also add a star layer to display stars in background (I use a catalogue from Hyparcos).

![Capture d'écran 2023-11-19 132744](https://github.com/maplibre/maplibre/assets/2014432/1f41583b-2f83-4871-a745-357c9e8206f5)

### Controls:
I do not change the control system and it work pretty well: moving, zooming or change the camera heading.

https://github.com/maplibre/maplibre/assets/2014432/d24c164a-2bd5-4a92-8f68-3a411c5f46c9


### TODO

**Performance:**
When the view is zoom at the mercator level, it could be better in term of performance to avoid the offscreen pass and to switch back to the default mercator rendering. For the moment, the offscreen pass is still used.

**Hole in pole:**
I did not manage to fill the hole after 85° of latitude.

![Capture d'écran 2023-11-19 133311](https://github.com/maplibre/maplibre/assets/2014432/3dfde07c-efd9-48f4-9af1-d2f35e936512)

**Terrain:**
As I use the terrain shader, I remove the elevation management. Still need to be link again.

**Symbols:**
All symbols remain in mercator projection. To avoid display problems in 3D globe, I implement a mechanism to hide symbols when zoom out and display it again when zoom in (after a configurable level).
The symbol position should be recomputed whith the appropriate placement and declutered to avoid superposition in 3D globe projection, and, of course, symbols behind the Earth should be hided (on tiles on borders of the Earth).

**Tile management:**
Improvement should be necessary to ensure all tiles are well loaded and displayed but also to ensure the right level is loaded according to the observer distance. For the moment, the same code as for mercator projection is used. This system load higher resolution tiles as necessary when the tiles is far from the observer.

**Atmosphere and stars:**
For the moment, the atmosphere is displayed for any map. It could be strange when its not the satellite view.

![Capture d'écran 2023-11-19 133023](https://github.com/maplibre/maplibre/assets/2014432/57443c12-6f0f-4b55-82ee-561b7ce10066)

**Other layers and features:**
All layer not managed by the offscreen pass should be rework to be projected into the globe view. I do not check all features also.

**Reply by HarelM** [reply]
Nice!!

**Reply by wipfli** [reply]
Oh wow, congratulations! Do you have a demo link?

**Reply by birkskyum** [reply]
or even better a pr :)

**Reply by sjg-wdw** [reply]
Neat!

**Reply by Pheonor** [reply]
Sorry, I just have a working version on my computer.

**Reply by kubapelc** [reply]
Hi,

I've also been working on globe for some time now, and have a working prototype of the render-to-texture (terrain inspired) approach as well as a vector approach. I've made a separate post in this discussion for details.

**Reply by Pheonor** [reply]
Hi,

I take the time to publish a demo here: https://pheonor.github.io/globe-demo/

**Reply by Pheonor** [reply]
You could notice something I am trying to fix: the sun is moving strangely when paning around the Earth ;)

**Reply by unphased** [reply]
Thanks for posting the demo. I did notice with this one the zoom level seems to randomly go in and out. It's quite jarring. Maybe you know what's causing it.

**Comment by kubapelc** [comment]
Hi,

I've also been working on globe for some time now, and have a working prototype of the render-to-texture (terrain inspired) approach, and I am now working on a prototype of rendering the vector tiles onto the globe "directly" - not using any intermediate framebuffers or textures, but projecting the polygons and lines onto the sphere in the vertex shader. My current goal is to evaluate which approach is better overall.

I have not tried to implement any custom controls, custom tile loading scheme or proper symbol positions and symbol hiding. I will do so once the main approach is decided.

The main advantage of the vector approach is that the resulting image remains pixel-perfect, regardless of screen resolution and other factors. There is no texture stretching or undersampling or oversampling. The vector approach may also be faster overall on mobile devices, since framebuffer switching and offscreen rendering in general is much slower there. This can already be a problem with the current terrain render-to-texture implementation, although mostly under very specific circumstances (looking at a complex spot under high pitch causes the texture cache to be insufficient and re-renders all tiles every frame - this is also slow on a PC). With the vector approach, no extra memory is needed for the texture cache.

Also, features with constant screenspace size will be rendered properly - think of the outline on highway lines, which preserve their thickness regardless of the zoom level. With render-to-texture approach, this would either not be the case or would require rerendering the cached textures every frame during movement, which will be noticably slow on mobile devices.

The main downside of the vector approach is that it is much harder to implement properly, that is without any single-pixel holes and other artifacts. It also contains an extra preprocessing step for the vector tiles, since all polygonal features must be subdivided enough for them to actually get projected to a curved shape (think of a triangle stretching from pole to pole - such geometry must be subdivided for the vertex shader to "round" it properly). This extra step can potentially slow down tile loading somewhat. This subdivision also increases the amount of geometry, although this does not seem to be an issue so far - the vector map is butter smooth on my 2019 midrange phone. I haven't looked at memory requirements, but assume that they will be significantly lower than storing a bunch of high resolution textures for the cached tiles in the render-to-texture approach. The subdivision is also mostly needed for low zoom levels (looking at the whole planet), higher zoom levels may not need any subdivision at all - I have yet to finetune this.

I am currently finishing the vector globe conversion for the "fill" layer type. I will go on to convert the "line" layer type next, and all other layer types after that should be significantly easier.

I've wanted to wait with unveiling this until the december MapLibre technical steering comittee meeting, and present my current results there, as well as ask about which of the two approaches they deem better for MapLibre. I figured it is better to post this right away.

Below is a screenshot of the vector globe, with layer types other than "fill" disabled.

![vectorglobe](https://github.com/maplibre/maplibre/assets/57600346/603b302b-9061-498c-bda6-38177d7c354a)

---

Demo: https://kvaleya.gitlab.io/maplibre/globe/globedemo.html

**Reply by mdorda** [reply]
@Pheonor Would you mind to make a call with us and joining forces?

**Reply by wipfli** [reply]
Exciting news, thanks for sharing @kubapelc

**Reply by HarelM** [reply]
Will it solve this kind of issues for example that we have in our 3D terrain - where the trail is draped on the terrain and doesn't look sharp?
![image](https://github.com/maplibre/maplibre/assets/3269297/b8911f53-02c2-406b-b7a8-ed1e315a3f80)
This is how this looks in 2D:
![image](https://github.com/maplibre/maplibre/assets/3269297/ea49ab60-b945-4698-bc9a-84b23f6548b9)

Refernce:
https://israelhiking.osm.org.il/map/15.63/32.8246/35.4967

**Reply by Pheonor** [reply]
@mdorda Ok for a call. I applied recently as MapLibre developer :)

**Reply by kubapelc** [reply]
> Will it solve this kind of issues for example that we have in our 3D terrain - where the trail is draped on the terrain and doesn't look sharp?

@HarelM It won't, I'm only implementing projection to a sphere and am not planning any modification to terrain rendering.

One could, in theory, use the subdivision logic I use to subdivide the vector tiles so that all primitives are cut on the grid of DEM pixels, and then deform the vector map by reading the DEM in the vertex shader. This would probably be fairly nontrivial to implement with correct occlusion and blending of the different layers. It would likely need a depth prepass and then doing depth testing manually in the pixel shader to avoid z-fighting.

**Reply by kubapelc** [reply]
Hello, here is a demo of the vector globe!

https://kvaleya.gitlab.io/maplibre/globe/globedemo.html

Some things to note:

- Only fill and line layers are enabled. Others aren't adapted for globe yet.
- I haven't yet touched any code related to map controls. It behaves the same way as if you were looking at a regular web mercator map.
- I haven't yet touched any code related to tile loading and determining which tiles to display. Tiles *will* appear and disappear at weird times in the demo.
- There are errors and distortion at high zoom levels. This is likely due to limited precision of trigonometry on the GPU and error accumulation when computing with 32bit floats. I plan on switching to plain old 2D rendering when zoom level is high enough for globe curvature to not be noticeable. This switch will happen long before these float errors of globe rendering become visible.

**Reply by HarelM** [reply]
Looks great! thanks for sharing!

**Reply by kubapelc** [reply]
I've updated the demo, symbols are now drawn on the globe. Only the vertex shader for symbols is modified to place them at the right position, and only basic SDF symbols are enabled (icons, etc. are disabled for now). I haven't changed symbol placement logic at all so far, so symbol visibility is still determined as if they were on a 2D web mercator map.

Tile preprocessing is slightly faster now (though it is still much slower than a flat map).

**Reply by birkskyum** [reply]
Very interesting, certainly helps to get the labels on. I noticed also some tearing of labels at high zoom
<img width="1172" alt="Screenshot 2023-12-13 at 14 51 40" src="https://github.com/maplibre/maplibre/assets/74932975/e561ebf2-4db9-4f1b-899d-1712fb959c52">

**Reply by kubapelc** [reply]
> Very interesting, certainly helps to get the labels on. I noticed also some tearing of labels at high zoom

Yeah, those are probably due to limited precision of floats or trigonometry on the GPU. Street names also get projected onto the globe, so they are also affected by this problem. I'm planning on transitioning to 2D mercator map that doesn't have precision problems at high zoom levels to hide these artifacts.

**Reply by kubapelc** [reply]
I've updated the demo with the latest globe version, which now supports raster tiles, hillshade, heatmaps and animated transtition to/from globe view. The demo also allows you to view several map types and switch between them as well as switch between globe view and mercator.

Globe view now automatically and seamlessly switches to mercator at high zoom levels to hide precision errors.

https://kvaleya.gitlab.io/maplibre/globe/globedemo.html

**Reply by birkskyum** [reply]
This looks amazing. I assume terrain is outside of scope for this initial globe - what do you find is still missing?

**Reply by kubapelc** [reply]
Vivian's globe implementation makes use of terrain's render-to-texture tiles already, so we currently plan to use his implementation for terrain :) Still missing from vector map is proper symbol placement/hiding, as it still assumes a mercator map, fill extrusion layer type, custom layers (not sure whether to even attempt those) and some styling options from other layers. I'm still only adapting the core rendering, so I haven't even looked at things like determining which tiles are visible on the globe (note how they still disappear randomly), user controls, etc.

**Reply by birkskyum** [reply]
I see. It's great progress still - good work!

**Reply by HarelM** [reply]
I think it will be smart to define style spec properties related to globe so we can agree on the style spec API together.

**Reply by birkskyum** [reply]
also, it would be great to make sure we can align this with deck.gl globe, so all our data layers keep working. @Pessimistress

**Reply by Pheonor** [reply]
Great Job !
The demo look very smooth.

**Reply by josuelmm** [reply]
Excelent Work

**Comment by Pheonor** [comment]
Hi,

I want to share with you some progress on the RTT globe/terrain implementation.

I port the POC on a pre 4 version of MapLibre (code available here: [https://github.com/Pheonor/maplibre-gl-js/tree/add-globe](https://github.com/Pheonor/maplibre-gl-js/tree/add-globe)).
I remove all the atmosphere and stars stuff to concentrate effort on Globe only.
I also remove the adaptative stuff.

For now, I improve the tile selection to avoid black area and too much tile loaded and not displayed with good effect on performance.
I still need to work on this tile selection to improve it, but perhaps on a second step.

Pole hole are filled by extending the tiles border to 90° and -90° but still remain some work to have a proper version.

Terrain is working. I setup a terrain for the whole Earth and it look good.
Still some glitch on the camera movement. I will try to fix that.

For the moment, I try to fix a precision problem when zooming.

Demo updated here: https://pheonor.github.io/globe-demo/

**Comment by Pheonor** [comment]
Hi,

As we are a first working version in the globe branch, I propose to open and work on three tickets:

    Tile loading improvement to better manage tile in globe mode
    Atmosphere layer
    Background color with stars

**Reply by HarelM** [reply]
- https://github.com/maplibre/maplibre-gl-js/issues/3887
- https://github.com/maplibre/maplibre-gl-js/issues/3888
- https://github.com/maplibre/maplibre-gl-js/issues/3889

**Comment by YohanSciubukgian** [comment]
Hi,

Will it be possible to have 3D models rendered (and aligned accordingly with the map) on the Globe View?

Could you consider having a documentation that could showcase an aircraft 3D model flying from JFK (New York) to GTW (London) airports on the both Globe view and when zooming-in, having the 3D model aligned on the 2D map accordingly?

Will it be possible to have the aircraft 3D model shadow rendered on the Globe View Map while moving from JFK to GTW airports?

**Reply by kubapelc** [reply]
Hi, I finally got around to it, you can see the PR which adds custom layer support including placing 3D models on the globe here: https://github.com/maplibre/maplibre-gl-js/pull/4577

It is not the aircraft example exactly, but the principle should be the same.

**Reply by YohanSciubukgian** [reply]
Your PR looks really promising !! Great jobs ! 🎉🎉
Do you know if shadow rendering of the 3D model is supported in your PR yet ?

**Comment by freeseus** [comment]
Just checking in... has there been any progress in adding a 3D globe map tiling option?

And, if indeed we did have a 3D globe tiling option for Maplibre in the future, would it be possible to transition from Globe to Mercator projection on the fly. Something like this (use the slider): https://jsfiddle.net/wgdtL0um/

**Reply by kubapelc** [reply]
Hi, I'm not sure what you mean by globe map tiling, but you can check out a demo of the globe here: https://kvaleya.gitlab.io/maplibre/globe/globedemo.html

Note the checkbox in the top left corner, which does something very similar to the animation you linked :)

**Reply by freeseus** [reply]
Using tile images in maplibre, like "images/tiles/{z}_{x}_{y}.webp"

**Comment by kubapelc** [comment]
New vector globe demo with globe-aware controls! https://kvaleya.gitlab.io/maplibre/globe/globedemo.html

**Reply by freeseus** [reply]
Great progress! Are there plans to change the "Tile loading & visibility calculations still assume a Mercator map, hence the disappearing tiles" ?

**Comment by wiesehahn** [comment]
Should this probably be listed as released in the roadmap? https://maplibre.org/roadmap/globe-view/

**Reply by louwers** [reply]
Yes.

**Comment by glaroc** [comment]
Should it be possible to add a pmtiles layer to the Globe view? I'm trying and not succeeding so was wondering if this is not implemented yet?

**Reply by HarelM** [reply]
PMTiles is not a layer, it's a source... I don't see a reason it won't work...

**Reply by glaroc** [reply]
Got it to work. I had to update the pmtiles library in React.

**Reply by fo-rk** [reply]
This is huge, thanks @kubapelc, @Pheonor and anybody else who contributed to this.

**Comment by joewdavies** [comment]
Thank you to all who contributed to this!

**Comment by geekdenz** [comment]
I see this is merged and the demos work. However, I was unable to find out how to configure this effect in my map.
Any hints how to configure this?

EDIT: Found out. You need to change your style.json file to include the projection property.

```json
  "projection": {
    "type": "globe"
  },
```

**Comment by stuneak** [comment]


**Reply by geekdenz** [reply]
Would it not be:

```javascript

		map.on('contextmenu', (data) => {
			const { point } = data;

			// Check if the point is on the globe surface
			const isOnGlobe = map.transform.isPointOnMapSurface(point);
		
		});
```

Is `newMap` a different map instance?

**Reply by stuneak** [reply]
Sorry, you're right. My bad :)

**Comment by geekdenz** [comment]
Is the map, when zoomed in in EPSG:3857 (web mercator) or still in the globe projection? Looks to me like Antarctica is in the globe projection to me but I could be mistaken. Would it be earthwide the most area preserving (the globe projection)?

**Reply by birkskyum** [reply]
It transition to web mercator when you zoom in, if it's set to just "globe" projection. It's worth noting that "globe" is a preset over our projections api. You can configure the break points for the transition from "vertical perspective" (a sphere) to "mercator" (web mercator), or disable the transition entirely by passing only vertical perspective.

For additional context, see
- https://maplibre.org/maplibre-style-spec/types/#projectiondefinition
- https://maplibre.org/maplibre-style-spec/projection/
